pkg/sys/drive_discovery_linux.go

```diff
@@ -738,43 +672,13 @@ func probeDevices() (devices map[string]*Device, err error) {
 	return devices, nil
 }
 
-func createDevice(event map[string]string) (device *Device, err error) {
-	name := filepath.Base(event["DEVPATH"])
-	if name == "" {
-		return nil, fmt.Errorf("event does not have valid DEVPATH %v", event["DEVPATH"])
-	}
-
-	major, err := strconv.Atoi(event["MAJOR"])
-	if err != nil {
-		return nil, err
-	}
-
-	minor, err := strconv.Atoi(event["MINOR"])
-	if err != nil {
-		return nil, err
-	}
-
-	switch event["ACTION"] {
-	case uevent.Add, uevent.Change:
-		// Older kernels like in CentOS 7 does not send all information about the device,
-		// hence read relevant data from /run/udev/data/b<major>:<minor>
-		info, err := readRunUdevData(major, minor)
-		if err != nil {
-			return nil, err
-		}
-		for key, value := range info {
-			if _, found := event[key]; !found {
-				event[key] = value
-			}
-		}
-	}
-
-	if device, err = newDevice(event, name, major, minor, strings.Contains(event["DEVPATH"], "/virtual/")); err != nil {
+func createDevice(udevData *udev.Data) (device *Device, err error) {
+	if device, err = newDevice(udevData, strings.Contains(udevData.Path, "/virtual/")); err != nil {
 		return nil, err
 	}
 
-	if event["ACTION"] == uevent.Remove {
-		return device, nil
+	if device.Name == "" {
```

Comment:
Move this error check at beginning of the function i.e. we fail early before `newDevice()`